src/System.Net.Mime/src/System/Net/Mail/SmtpClient.cs

```diff
@@ -0,0 +1,1135 @@
+        }
+
+        [HostProtection(ExternalThreading = true)]
+        public void SendAsync(MailMessage message, object userToken)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(this.GetType().FullName);
+            }
+
+            if (NetEventSource.Log.IsEnabled())
+            {
+                NetEventSource.Enter(NetEventSource.ComponentType.Web, this, "SendAsync", "DeliveryMethod=" + DeliveryMethod.ToString());
+            }
+
+            if (GlobalLog.IsEnabled)
+            {
+                GlobalLog.Enter("SmtpClient#" + LoggingHash.HashString(this) + "::SendAsync Transport#" + LoggingHash.HashString(_transport));
+            }
+
+            try
+            {
+                if (InCall)
+                {
+                    throw new InvalidOperationException(SR.net_inasync);
+                }
+
+                if (message == null)
+                {
+                    throw new ArgumentNullException("message");
+                }
+
+                if (DeliveryMethod == SmtpDeliveryMethod.Network)
+                    CheckHostAndPort();
+
+                _recipients = new MailAddressCollection();
+
+                if (message.From == null)
+                {
+                    throw new InvalidOperationException(SR.SmtpFromRequired);
+                }
+
+                if (message.To != null)
+                {
+                    foreach (MailAddress address in message.To)
+                    {
+                        _recipients.Add(address);
+                    }
+                }
+                if (message.Bcc != null)
+                {
+                    foreach (MailAddress address in message.Bcc)
+                    {
+                        _recipients.Add(address);
+                    }
+                }
+                if (message.CC != null)
+                {
+                    foreach (MailAddress address in message.CC)
+                    {
+                        _recipients.Add(address);
+                    }
+                }
+
+                if (_recipients.Count == 0)
+                {
+                    throw new InvalidOperationException(SR.SmtpRecipientRequired);
+                }
+
+                try
+                {
+                    InCall = true;
+                    _cancelled = false;
+                    _message = message;
+                    string pickupDirectory = PickupDirectoryLocation;
+
+                    CredentialCache cache;
+                    // Skip token capturing if no credentials are used or they don't include a default one.
+                    // Also do capture the token if ICredential is not of CredentialCache type so we don't know what the exact credential response will be.
+                    _transport.IdentityRequired = Credentials != null && (Credentials is SystemNetworkCredential || (cache = Credentials as CredentialCache) == null);
```

Comment:
If `Credentials` is `CredentialCache`, the desktop code also checks to see if the `SystemNetworkCredential` is in the cache via the internal `CredentialCache.IsDefaultInCache` property. `CredentialCache` in .NET Core no longer tracks that (and the internal property wouldn't be accessible here anyway). We might need to open a tracking bug to determine how to address this.